src/pytorch_lightning/strategies/colossalai.py

```diff
@@ -0,0 +1,294 @@
+        is_training = self.lightning_module.trainer and self.lightning_module.trainer.training
+        if is_training:
+            if len(self.optimizers) > 1:
+                raise MisconfigurationException("ColossalAIStrategy only supports single Optimizer now.")
+            optimizer = self.optimizers[0]
+            assert isinstance(optimizer, (CPUAdam, HybridAdam)), (
+                "ColossalAIStrategy only supports colossalai.nn.optimizer.CPUAdam "
+                "and colossalai.nn.optimizer.HybridAdam."
+            )
+        pl_module = self.model
+        if not hasattr(pl_module, "_colossalai_zero"):
+            if self.use_chunk:
+                chunk_size = self.chunk_size or ChunkManager.search_chunk_size(
+                    self.model, **self.chunk_size_search_kwargs
+                )
+            else:
+                chunk_size = None
+            chunk_manager = ChunkManager(
+                chunk_size,
+                self.process_group,
+                self.enable_distributed_storage,
+                GeminiManager.get_default_device(self.placement_policy),
+            )
+            gemini_manager = GeminiManager(self.placement_policy, chunk_manager)
+            assert isinstance(self.model, (pl.LightningModule, _LightningPrecisionModuleWrapperBase))
+            model = _LightningModuleWrapperBase(self.model)
+            self.model = ZeroDDP(model, gemini_manager, self.force_outputs_fp32)
+            pl_module._colossalai_zero = [self.model]
+        else:
+            self.model = pl_module._colossalai_zero[0]
+        if is_training:
+            self.optimizers = [
+                ZeroOptimizer(optimizer, self.model, gpu_margin_mem_ratio=self.gpu_margin_mem_ratio, **self.amp_kwargs)
+            ]
+
+    def setup(self, trainer: "pl.Trainer") -> None:
+        if not isinstance(self.accelerator, CUDAAccelerator):
+            raise MisconfigurationException(
+                f"Colossalai is only supported on GPU but `{self.accelerator.__class__.__name__}` is used."
+            )
+        self.accelerator.setup(trainer)
+
+        assert self.lightning_module is not None
+        self.lightning_module._device = self.root_device
+
+        if trainer.accumulate_grad_batches > 1:
+            raise MisconfigurationException(
+                "Colossalai does not support gradient accumulation now. " "please set `accumulate_grad_batches` to 1."
+            )
+
+        accumulation_scheduler = trainer.accumulation_scheduler
+        if accumulation_scheduler.epochs != [0]:
+            raise MisconfigurationException(
+                "Colossalai currently does not support different `accumulate_grad_batches` at different epochs."
+            )
+        self.setup_optimizers(trainer)
+
+        if not isinstance(self.precision_plugin, ColossalAIPrecisionPlugin):
+            raise MisconfigurationException("`ColossalaiStrategy` is only compatible with `ColossalAIPrecisionPlugin`.")
```

Comment:
can you move all these exceptions at the top in this method?

Reply:
Done.